### PR TITLE
Add sub-reader for PHI Versaprobe data

### PR DESCRIPTION
**Pull request by lukaspie**
Phi Parser:
- Parser for PHI Versaprobe .spe (regular spectra) and .pro (sputter depth profiling)
- New config file to account for these data sets
- Correct parsing of binary encoded data
- Mapping of optional instrument devices (sputter source, neutral ion gun, C60 gun, defect positioner) to the config file

Other changes:
- The main reader supports energy reading as well as more options for writing raw detector data (i.e., write scan data if there is no data for the individual channels

Error fixes:
- Links that are broken (see #2) are temporarily removed from config files until that bug is fixed.
- The mapping of keys and the renaming of `ENTRY[entry]` to the actual entries (e.g. `ENTRY[MeasurementGroup_Survey]) has been streamlined and harmonized across all subreaders

Still missing:
- [ ] Example data for this reader (pending approval by data provider)
- [ ] Test data for this reader (pending approval by data provider)

This PR depends on FAIRmat-NFDI/pynxtools#220 and should be merged afterwards/together.
EDIT: FAIRmat-NFDI/pynxtools#220 merged into pynxools-xps@main.

Resolves FAIRmat-NFDI/pynxtools-xps#1.

### CODE REVIEW

File: examples/phi/README.md
Comment (domna) [inline examples/phi/README.md:3]:
```suggestion
This is an example for [Phi MultiPak](https://www.phi.com/surface-analysis-equipment/genesis.html#software:multi-pak-data-reduction-software/) .spe (single spectra) and .pro (profile / external parameter / ....) files, which is the propietary format of PHI Electronics used for their VersaProbe 4 instrumens. The Phi MultiPak software version that was used to measure this data is SS 3.3.3.2.1. The example conversion can be run with the following commands.
```

File: pynxtools_xps/phi/spe_pro_phi.py
Comment (domna) [inline pynxtools_xps/phi/spe_pro_phi.py:276]: Should this not be coming from the config file?

File: pynxtools_xps/phi/spe_pro_phi.py
Comment (domna) [inline pynxtools_xps/phi/spe_pro_phi.py:727]: Maybe a bit cleaner and w/o slow try...except + it's clearer what is actually checked:
```suggestion
                mpes_key = spectrum_key
                self._xps_dict[f"{root}/{mpes_key}"] = spectrum.get(spectrum_key)

```

that way the key would be filled with `None` if the key does not exists. If this is not fine maybe like this
```suggestion
                mpes_key = spectrum_key
                if spectrum_key in spectrum:
                    self._xps_dict[f"{root}/{mpes_key}"] = spectrum[spectrum_key]

```

File: pynxtools_xps/phi/spe_pro_phi.py
Comment (domna) [inline pynxtools_xps/phi/spe_pro_phi.py:861]: I'm not entirely sure but I think there is a function in python with which you can determine whether the file was saved in windows or unix encoding line endings.

File: pynxtools_xps/phi/spe_pro_phi.py
Comment (domna) [inline pynxtools_xps/phi/spe_pro_phi.py:902]: Consider making this a dict and looping over the dict here

File: pynxtools_xps/reader_utils.py
Comment (domna) [inline pynxtools_xps/reader_utils.py:289]:
```suggestion
        scan_key = "scans/Scan_0"
```

Comment (domna) [reply pynxtools_xps/reader_utils.py:289]: Nothing is inserted here, so the f-string is not necessary

File: pynxtools_xps/reader_utils.py
Comment (domna) [inline pynxtools_xps/reader_utils.py:306]: Consider using `re.compile`. Maybe even outside the function if this is run many times.

**Review by domna — CHANGES_REQUESTED** [review]
There are some small things and my general confusion about the dict (whether this can be a config file). But probably there is a good reason this is separate from the config file which I just don't understand.

I tested it with the files from Sebastian and it all works well.

An additional consideration we already discussed in discord is adding an additional axis for the depth profile so you can view it in one NXdata. Depending on the effort this can either be handled here or in a separate PR.

Thanks for the implementation :)

File: examples/phi/README.md
Comment (lukaspie) [reply examples/phi/README.md:3]: Thanks, I changed it

File: pynxtools_xps/phi/spe_pro_phi.py
Comment (lukaspie) [reply pynxtools_xps/phi/spe_pro_phi.py:727]: I went with the first suggestion, that works.

File: pynxtools_xps/reader_utils.py
Comment (lukaspie) [reply pynxtools_xps/reader_utils.py:289]: removed

File: pynxtools_xps/reader_utils.py
Comment (lukaspie) [reply pynxtools_xps/reader_utils.py]: done

File: pynxtools_xps/phi/spe_pro_phi.py
Comment (lukaspie) [reply pynxtools_xps/phi/spe_pro_phi.py:858]: I think it is only possible with [chardet](https://pypi.org/project/chardet/). So, in order to keep the dependencies small, I would keep it as it is.

File: pynxtools_xps/phi/spe_pro_phi.py
Comment (lukaspie) [reply pynxtools_xps/phi/spe_pro_phi.py:276]: This comes before the config file is even parsed. I generally use a `mapper` layer that is called by the general XPS reader and acts as an interface to different sub-parsers (i.e., for different software versions). This mapper then creates from the parser data a dictionary. This dict is used by the main parser to fill the template through the config file (with a token, similarly to what you have in the multi-format reader). The advantage of having this mapper layer is that I get to keep some functionality across the different file formats, so that the config file and filling of the template are always relatively similar. See also my comment [here](https://github.com/FAIRmat-NFDI/pynxtools/pull/250#issuecomment-1959742939).

It is a bit inefficient because I first parse the data (on the parser level) in a list of dicts (one for each spectrum) and then I populate the large mapper dict from there (though the performance is not that bad). In general, the parsers should create such dicts immediately. Eventually, I want to change it (see issue #19), but I will only do it once I switch to the multi-format reader.

Comment (domna) [reply pynxtools_xps/phi/spe_pro_phi.py:276]: Ok, makes sense. I was just a bit confused when I read it but I think it's totally fine. So I'll just resolve this

File: pynxtools_xps/phi/spe_pro_phi.py
Comment (lukaspie) [reply pynxtools_xps/phi/spe_pro_phi.py]: done

**Comment by lukaspie** [comment]
> An additional consideration we already discussed in discord is adding an additional axis for the depth profile so you can view it in one NXdata. Depending on the effort this can either be handled here or in a separate PR.

I realized that this is a bit of an involved effort because the XPS reader so far only works for one-dimensional datasets with an energy axis. In principle, I think I should redesign it such that not every scan gets an axis in NXdata, but rather the scan numbers, measurement cycle numbers, sputter cycle numbers, are axes to one NXdata. However, this requires quite a bit of effort to disentangle in the main reader, so I suggest that I will do it in a separate PR. 

Thus, I would just simply wait for the approval that the example/test data can be included here and then it would merge this PR.

**Comment by domna** [comment]
> Thus, I would just simply wait for the approval that the example/test data can be included here and then it would merge this PR.

Sounds reasonable. I'll approve this already. Feel free to add the test data and merge when ready

**Review by domna — APPROVED** [review]

**Pull request merged**
